aten/src/ATen/native/cuda/PersistentSoftmax.cuh

```diff
@@ -0,0 +1,385 @@
+            if (element_index < element_count) {
+                if (is_log_softmax) {
+                    dst[i*element_count+it*WARP_SIZE] = elements[i][it] - sum[i];
+                } else {
+                    dst[i*element_count+it*WARP_SIZE] = elements[i][it] / sum[i];
+                }
+            } else {
+                break;
+            }
+        }
+    }
+}
+
+
+// WARP_BATCH number of batches.
+// WARP_ITERATOINS The number of iterations required for one warp to iterate over all data.
+// WARP_SIZE number of elements working on a single batch, has to be a power of two.
+template <typename input_t, typename output_t>
+using softmax_forward_func = void(*)(output_t *dst, const input_t *src, int batch_size, int stride, int element_count);
+
+template <typename input_t, typename output_t, typename acc_t, bool is_log_softmax>
+bool warp_softmax_kernel(int log2_elements, int &warp_size, int &batches_per_warp, softmax_forward_func<input_t, output_t> &kernel) {
+    // determine size of a warp
+    const int next_power_of_two = 1 << log2_elements;
+    warp_size = (next_power_of_two < 32) ? next_power_of_two : 32;
+
+    // determine how many batches a warp should process.
+    batches_per_warp = (next_power_of_two <= 128) ? 2 : 1;
+
+    switch (log2_elements) {
+    case 0: // 1
+        kernel = &softmax_warp_forward<input_t, output_t, acc_t, 2,1,1, is_log_softmax>;
+        break;
+    case 1: // 2
+        kernel = &softmax_warp_forward<input_t, output_t, acc_t, 2,1,2, is_log_softmax>;
+        break;
+    case 2: // 4
+        kernel = &softmax_warp_forward<input_t, output_t, acc_t, 2,1,4, is_log_softmax>;
+        break;
+    case 3: // 8
+        kernel = &softmax_warp_forward<input_t, output_t, acc_t, 2,1,8, is_log_softmax>;
+        break;
+    case 4: // 16
+        kernel = &softmax_warp_forward<input_t, output_t, acc_t, 2,1,16, is_log_softmax>;
+        break;
+    case 5: // 32
+        kernel = &softmax_warp_forward<input_t, output_t, acc_t, 2,1,32, is_log_softmax>;
+        break;
+    case 6: // 64
+        kernel = &softmax_warp_forward<input_t, output_t, acc_t, 2,2,32, is_log_softmax>;
+        break;
+    case 7: // 128
+        kernel = &softmax_warp_forward<input_t, output_t, acc_t, 2,4,32, is_log_softmax>;
+        break;
+    case 8: // 256
+        kernel = &softmax_warp_forward<input_t, output_t, acc_t, 1,8,32, is_log_softmax>;
+        break;
+    case 9: // 512
+        kernel = &softmax_warp_forward<input_t, output_t, acc_t, 1,16,32, is_log_softmax>;
+        break;
+    case 10: // 1024
+        kernel = &softmax_warp_forward<input_t, output_t, acc_t, 1,32,32, is_log_softmax>;
+        break;
+    default:
+        return false;
+    }
+    return true;
+}
+
+template<typename input_t, typename output_t, typename acc_t, bool is_log_softmax>
+bool dispatch_softmax(output_t *dst, const input_t *src, int softmax_elements, int softmax_elements_stride, int batch_count)
```

Comment:
return value of dispatch_softmax is never checked, so you might as well make it void.

Reply:
I think I prefer to assert on the return value instead. dispatch_softmax will return false if you call it for a case that it can't handle properly.